driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
> One is even if ctx.Done() happens and no matter it calls netConn.Close() or netConn.SetDeadline(), mysqlConn.pingImpl() may still send nil error to line 65 before these two functions interrupt it. In this situation, netConn doesn't have to be closed and it's safe to reuse this connection.

Cancelleration is rare situation.  Code simplicity (maintenance cost) is more important than how cancellation efficient.

> The another reason is about the life cycle of mysqlConn. I think It is created and managed in database/sql/sql.go. It should be destroyed in the same level/place.

But TCP lifetime can be shorter than mysqlConn's lifetime.